[Xiu Wen] iP

YoYoCiti opened the pull request:
### **DukePro**
> "Your mind is for having ideas, not holding them." - David Allen

DukePro frees your mind of having to remember things you need to do. It's, 
* text-based
* easy to learn
*  ~FAST~ SUPER FAST to use

All you need to do is,
1. Download it from [here](https://github.com/YoYoCiti/ip/releases)
2. double click it
3. add your tasks
4. let it manage your tasks for you 😆 

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines 
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```javascript
public static void main(String[] args) {
    String projectDir = System.getProperty("user.dir");
    Path path = Paths.get(projectDir, "data", "duke.txt");
    new Duke(path).run();
}
```

Commit:
Add Gradle support

YoYoCiti committed:
add greet, echo and exit feature

YoYoCiti committed:
add store text in list and display list feature

YoYoCiti committed:
tasks are stored in list and add ability to mark tasks as done

YoYoCiti committed:
add support for tracking todos, events and deadlines

YoYoCiti committed:
update test files for automated testing

YoYoCiti committed:
add incorrect inputs error handling

YoYoCiti committed:
add delete tasks from list feature

YoYoCiti committed:
add enumerations and improved error handling

YoYoCiti committed:
Add feature to save tasks in hard disk and load them when Duke starts up

YoYoCiti committed:
Add date and time functionality

YoYoCiti committed:
Merge branch 'branch-Level-7'

YoYoCiti committed:
Merge branch 'branch-Level-8'

YoYoCiti committed:
Fix date/time parsing error after merge

YoYoCiti committed:
Refactor code to extract Ui, Storage, Parser and TaskList classes

YoYoCiti committed:
Organize classes into packages

YoYoCiti committed:
Add JUnit tests for Parser and TaskList

YoYoCiti committed:
Merge branch 'add-gradle-support' of https://github.com/YoYoCiti/ip

YoYoCiti committed:
Package app as JAR file

YoYoCiti committed:
Add JavaDoc comments

YoYoCiti committed:
Modify code to comply with coding standard

YoYoCiti committed:
Add feature to find task by searching for keyword

YoYoCiti committed:
Merge branch 'branch-A-JavaDoc'

YoYoCiti committed:
Merge branch 'branch-A-CodingStandard'

YoYoCiti committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/parser/Parser.java

YoYoCiti committed:
Add checkstyle configurations

YoYoCiti committed:
Modify code to adhere to coding standards

YoYoCiti committed:
Merge branch 'branch-A-Checkstyle'

YoYoCiti committed:
Add Duke GUI

YoYoCiti committed:
Merge branch 'branch-Level-10'

waynetee commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,67 @@
+            Commands cmd = Commands.valueOf(
+                    i == -1 
+                    ? fullCommand.toUpperCase() 
+                    : fullCommand.substring(0, i).toUpperCase());
```
How about you use the `String.split` function, would that make it cleaner?

waynetee commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,101 @@
+    public void saveData(ArrayList<Task> lst) {
+        try {
+            BufferedWriter bf = new BufferedWriter(new FileWriter(this.filePath.toFile()));
+            Task[] toIterate = lst.toArray(new Task[0]);
```
Why not iterate through `lst` directly instead of creating a new array?

waynetee commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,101 @@
+                boolean isDone = txt[1].equals("1");
+                switch (txt[0]) {
+                case "T":
+                    lst.add(new Todo(txt[2], isDone));
```
Would it be better to define a constructor that either takes in the whole string or takes in the individual parameters?

waynetee commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,48 @@
+     * @param msg the message to print
+     * @return given message
+     */
+    public String print(String msg) {
+        return msg;
+    }
```
What is the purpose of this method?

tanhuakun replied:
I am interested in the purpose of this method too.

waynetee reviewed (commented):
Hi Xiu Wen, I have added my comments, do let me know if you want to clarify anything. Thanks!

tanhuakun commented on src/test/java/duke/data/TaskListTest.java:
```diff
@@ -0,0 +1,54 @@
+
+public class TaskListTest {
+    @Test
+    public void testAddTask() {
```
It is not necessary to add the word "test" in your test methods. addTask() is sufficient in this case.

tanhuakun commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,34 @@
+    private String filter;
+
+    /**
+     * Constructor for FindCommand
+     * @param rest the user input after the command
+     * @throws EmptyDescriptionException if input is empty
+     */
```
Javadoc should have an empty line between [description and parameter section](https://se-education.org/guides/conventions/java/intermediate.html#comments). I have noticed this issue in many other files too. 
```suggestion
     * Constructor for FindCommand
     *
     * @param rest the user input after the command
     * @throws EmptyDescriptionException if input is empty
     */
```

tanhuakun commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,31 @@
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        storage.saveData(tasks.add(this.t));
+        return ui.print(String.format(
+                        "Got it. I've added this task:\n %s\nNow you have %d task(s) in the list",
+                        this.t.toString(),
+                        tasks.getCount()));
```
The indentation here might be too big. Indentation for wrapped lines [should be 8 spaces more than the parent line.](https://se-education.org/guides/conventions/java/intermediate.html#layout)

tanhuakun reviewed (commented):
LGTM. Overall good quality in code, with just a few coding standard violations to fix.

YoYoCiti committed:
add invalid index format exception

YoYoCiti committed:
update main class to Launcher for build gradle file

YoYoCiti committed:
Extract date time format pattern into static variable in Task

Deadline, Event and InvalidTimeException classes have common date and
time format patterns which is inputted as literals.

Common date and time format patterns causes code duplication as well as
use of magic literals.

Extracting the date time format pattern into a static variable stored in
the Task class prevents the use of magic literals. Format pattern can
also be changed conveniently without having to look for instances in
the code where it is used.

YoYoCiti committed:
Extract boolean expression in parse method into variable

The purpose of existing boolean expression i == -1 is unclear to
readers, especially if unfamiliar to indexOf method.

Extracting the expression into boolean variable allows readers to
understand the purpose of the boolean expression, which is to check
if the user input is a single word.

YoYoCiti committed:
Refactor code to improve code quality

YoYoCiti committed:
Use streams to print task list and find tasks

YoYoCiti committed:
Add assertions to document assumption that data file should be
written and read in the correct format

YoYoCiti committed:
Merge pull request #3 from YoYoCiti/branch-A-Streams

Use streams to print task list and find tasks

YoYoCiti committed:
Merge branch 'master' into branch-A-Assertions

YoYoCiti committed:
Merge branch 'master' into branch-A-CodeQuality

YoYoCiti committed:
Merge pull request #2 from YoYoCiti/branch-A-CodeQuality

Branch a code quality

YoYoCiti committed:
Merge branch 'master' into branch-A-Assertions

YoYoCiti committed:
Merge pull request #1 from YoYoCiti/branch-A-Assertions

Add assertions to document assumption that data file should be

YoYoCiti committed:
Adjust dialog box min height so that all text can be displayed

YoYoCiti committed:
Detect duplicate tasks and prevent user from adding them

YoYoCiti committed:
Merge branch 'branch-C-DetectDuplicates'

YoYoCiti committed:
Add yml file for CI

YoYoCiti committed:
Delete gradle.yml file

YoYoCiti committed:
Create gradle.yml

YoYoCiti committed:
Update gradle.yml

YoYoCiti committed:
Update gradle.yml

YoYoCiti committed:
Update gradle.yml to run tests instead of checks

YoYoCiti committed:
Merge branch 'master' into branch-A-BetterGui

YoYoCiti committed:
Add text box and background color into gui

YoYoCiti committed:
Merge branch 'branch-A-BetterGui'

YoYoCiti committed:
Add help command

YoYoCiti committed:
Adjust main window dimensions

YoYoCiti committed:
Modify find command to be case insensitive

YoYoCiti committed:
Add screenshot of product to docs folder

YoYoCiti committed:
Update README.md

YoYoCiti committed:
Update README.md to include latest release file

YoYoCiti committed:
Update gradle.yml to include checks

YoYoCiti committed:
Fix code to pass checkstyle test

YoYoCiti committed:
Merge branch 'master' of https://github.com/YoYoCiti/ip
